### ISSUE TITLE
investigate integration of Widoco metadata with ORR

### ISSUE THREAD

**Issue by graybeal** [mmisw/orr-portal#130]
Widoco (http://dgarijo.github.io/Widoco/) seems to be a very nice documentation tool for ontologies, and their recommendations (http://dgarijo.github.io/Widoco/doc/bestPractices/index-en.html) seem to be nice too (though MOD attribute list is more comprehensive: http://www.lirmm.fr/~jonquet/publications/documents/Article_MTSR-2017_MOD1.2.pdf). 

The goal is to evaluate the adoption of Widoco as an ORR-supported documentation tool, that could present to end users additional views documenting their ontologies. I do not have any idea whether this is even realistic (as Widoco seems to be a standalone tool) but I wanted to capture the possibility.

This would require adopting a compatible metadata standard for use within the ontology documentation as well, I expect.

**Comment by carueda** [comment]
@graybeal, @lewismc Just marked this as 'good first issue' especially given its "evaluation" nature. Please chime in if you have any reactions to this. Thx!

**Comment by lewismc** [comment]
WIDOCO itself is pretty cool however there are two things wrong here IMHO
- what happens if the new comer doesn't even really know what an ontology is? Say they just want to develop and improve the portal?
- WIDOCO is a standalone application... meaning that currently we need to download and install it, currently this makes it unusable in COR. We 'could' however work with the WIDOCO team to have the [artifacts published to Maven Central](https://github.com/dgarijo/Widoco/issues/301).

**Comment by graybeal** [comment]
> what happens if the new comer doesn't even really know what an ontology is? Say they just want to develop and improve the portal?

As with all newcomers (and existing developers), their work is evaluated as they present it. Even without any knowledge of ontologies, I could make a reasoned pass at how to include or reference an external system like WIDOCO,  demonstrate it in a branch on my own machine (caveats below re complex integration), and identify issues and benefits that I see. Then the people evaluating whether to incorporate my work could comment and determine whether it is heading in an acceptable direction, or what would need to change.

> WIDOCO is a standalone application... meaning that currently we need to download and install it, currently this makes it unusable in COR.  

I don't know how to interpret this. It is usable by COR if the developer downloads and installs it on the COR machine—or on another machine—from which COR could run the CLI, yes? Or does it require operation from a UI? (Sorry, if I had more time I'd go look this up, I'm being bad.)  I know Carlos doesn't like having more software running on the COR machine, and if we agree that makes sense, WIDOCO may be a non-starter—or maybe could be run as a separate hosted system. Somewhere.

> We 'could' however work with the WIDOCO team to have the artifacts published to Maven Central. 

What is the advantage of having it in Maven Central, as opposed to GitHub?  (I'm assuming you mean code artifacts, not the actual WIDOCO outputs.)

> Please chime in if you have any reactions to this. 

I think there is likely to be a significant integration challenge, in that a lot of different artifacts get produced and then must be accessed/displayed from COR.  However, if the first task is framed as performing the analysis—testing the system and evaluating how it could be integrated with COR, and writing up how hard that would be—then it is a low-risk thing for someone to do.

But  there are significant other tasks that I think would be needed. In particular, the COR metadata needs to be aligned with this, and (IMHO) with the MOD metadata recommendations (I think that's in another task?). And that strongly suggests that we should communicate with the WIDOCO team to see if they are willing to get aligned with MOD recommendations.

**Comment by carueda** [comment]
Agree pretty much with John, especially re the evaluation nature of this entry as I already mentioned.

(oh, just a precision: it should be just OK to have "more software running on the COR machine" as long as it has the relevant compute/space/admin resources.)

**Comment by kunakl07** [comment]
@lewismc 
Sir, I would like to work on this issue

**Comment by lewismc** [comment]
@kunakl07 OK, so have you built ORR yet using the [Docker deployment](https://github.com/mmisw/orr-instance-template)?

If so then I advise you to start looking at editing the portal code to implement some features. The issue is pretty well described above. Do you need more guidance/ If so please explain what it is that you need guidance on. Thanks

**Comment by lewismc** [comment]
Hi @graybeal I had previously not built Widoco or used it... I just did both and would like to start working on this issue but there are a few things we need to iron out. Rather than going through what each one of them is, I am instead going to propose what I think we should do. 

**Current Resource Page Layout**
Very simply put, a resource metadata window is positioned above a tabular data window which offers various ways for interacting with the data. 
![imageedit_2_6273091994](https://user-images.githubusercontent.com/1165719/65559633-53259500-def0-11e9-98f9-81546ed2ab96.png)

**Proposed Resource Page Layout**
All of the following metadata should be captured in the Widoco representation
* general ontology metadata: rdfs:label, rdfs:comment, owl:versionInfo
* usage/license/permissions: omvm:Manager of source vocabulary, omvm:Contact/Responsible party, omvm:Contact role, omvm:Author credit required and omvm:Citation string
* Original source...
* Other ...
... therefore why do we need an entire window only to duplicate it?

<img width="1528" alt="Screen Shot 2019-09-24 at 5 36 02 PM" src="https://user-images.githubusercontent.com/1165719/65560006-d09dd500-def1-11e9-86f6-239395e15ae7.png">

Embedding Widoco in this page would also remove the bottom data panel as well by providing all necessary information and vizualization. The following shows embedded WebVOWL as well

<img width="1680" alt="Screen Shot 2019-09-24 at 5 42 01 PM" src="https://user-images.githubusercontent.com/1165719/65560273-d2b46380-def2-11e9-8c73-6decd0985a0a.png">

What do you think about the proposal above? 

In summary here are some of the selling points

* reduces orr-portal attempting to reproduce metadata fields which are not present... and needlessly wasting real estate in their absence. 
* removes the currently defunct LODE button replacing all of the functionality in the embedded Widoco frame
* provides more appealing data representation than we already had
* replicates the linked data navigation functionality where all IRI's are still hyperlinked and can be navigated

**Comment by lewismc** [comment]
I traced the code which pulls in [appConfig.externalTools.ontViewers](https://github.com/mmisw/orr-portal/blob/master/src/app/js/config.js#L68-L79) to [ont-data.js](https://github.com/mmisw/orr-portal/blob/master/src/app/js/ont/ont-data.js) which is then rendered in the [ont-data.html](https://github.com/mmisw/orr-portal/blob/master/src/app/js/ont/ont-data.html) template. 

I don't think it would be much work to simplify the deployment but I am not very familiar with Angularjs so would need to step through this with @carueda 

The primary question I have right now is whether we would want to run Widoco ad-hoc e.g. whenever a page is rendered... or else pre-generate all of the outputs. Based on peformance concerns I had using Widoco locally I would prefer we pre-generate everything but this introduces some new questions
* how do we run it for all existing resources?
* where is this stored on the Docker container hosting orr-portal
* how do we route this in the application
* how do we invoke this pipeline upon new data or data changes within the wider system?

The above questions become non-trivial and would take quite a bit of additional work. Can you guys think of anything else?

**Comment by lewismc** [comment]
I propose to close this issue off and work together on implementing the plugin for pyLODE https://github.com/RDFLib/pyLODE
If there are objections @graybeal then please re-open.

**Comment by carueda** [comment]
@lewismc Sorry, I realized I lost track of this one and missed to note your good points.  But I would agree we could revisit them in the context of enabling pyLODE.  (In any case, the Widoco system seems very appealing as well, so we should keep it in mind anyway, I think.)
